### PR TITLE
OSDOCS#6030: cert-manager: setting log levels

### PR DESCRIPTION
**Pull request by xenolinux**
Version(s): 4.13+

Issue: https://issues.redhat.com/browse/OSDOCS-6030

Link to docs preview: https://59777--docspreview.netlify.app/openshift-enterprise/latest/security/cert_manager_operator/cert-manager-log-levels.html

QE review:
- [x] QE has approved this change.

SME review:
- [x] SME has approved this change.

Additional information:

### CODE REVIEW

File: modules/cert-manager-enable-operator-log-level.adoc [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+You can set a log level for the {cert-manager-operator} to determine the amount of detail in logged error messages. The valid values for the {cert-manager-operator} log level are `Normal`, `Debug`, `Trace`, and `TraceAll`.
+
+.Prerequisites
+
+* You have access to the cluster with `cluster-admin` privileges.
+* The {cert-manager-operator} is installed.
+
+.Procedure
+
+. Update the deployment for the {cert-manager-operator} to use the config map by running the following command:
```
Comment (swghosh) [inline modules/cert-manager-enable-operator-log-level.adoc:18]: Would prefer removing `config map` as we're just patching the subscription object and making no changes to any config map at this step.

Comment (swghosh) [reply modules/cert-manager-enable-operator-log-level.adoc:18]: How about something like: `Update the subscription object specific to {cert-manager-operator} in order to provide the verbosity level for the operator logs:`

File: modules/cert-manager-enable-operator-log-level.adoc [hunk 2]
```diff
@@ -0,0 +1,34 @@
+// Module included in the following assemblies:
+//
+// * security/cert_manager_operator/cert-manager-log-levels.adoc
+
+:_content-type: PROCEDURE
+[id="cert-manager-enable-operator-log-level_{context}"]
+= Setting a log level for the {cert-manager-operator}
+
+You can set a log level for the {cert-manager-operator} to determine the amount of detail in logged error messages. The valid values for the {cert-manager-operator} log level are `Normal`, `Debug`, `Trace`, and `TraceAll`.
```
[the remark this answers is not in the document]
Comment (swghosh) [reply modules/cert-manager-enable-operator-log-level.adoc:9]: Also would suggest, removing: `The valid values for the {cert-manager-operator} log level are Normal, Debug, Trace, and TraceAll.` as for operator log level we're just using numeric verbosity and not these enum values.

File: modules/cert-manager-enable-operand-log-level.adoc [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+//
+// * security/cert_manager_operator/cert-manager-log-levels.adoc
+
+:_content-type: PROCEDURE
+[id="cert-manager-enable-operand-log-level_{context}"]
+= Setting a log level for the cert-manager controller
+
+You can set a log level for the cert-manager controller to determine the amount of detail in logged error messages.
+
+[NOTE]
```
[the remark this answers is not in the document]
Comment (swghosh) [reply modules/cert-manager-enable-operand-log-level.adoc:11]: Probably, if we want to add such a section, it'd be good to put descriptions of each log level type via details mentioned in https://github.com/openshift/api/blob/894b49f57a15cbce3869961e20cd9d52df6f8b0f/operator/v1/types.go#L97-L110

File: modules/cert-manager-enable-operator-log-level.adoc [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+* You have access to the cluster with `cluster-admin` privileges.
+* The {cert-manager-operator} is installed.
+
+.Procedure
+
+. Update the subscription object for {cert-manager-operator} to provide the verbosity level for the operator logs by running the following command:
++
+[source,terminal]
+----
+$ oc -n cert-manager-operator patch subscription cert-manager-operator --type='merge' -p '{"spec":{"config":{"env":[{"name":"OPERATOR_LOG_LEVEL","value":"v"}]}}}' <1>
```
[the remark this answers is not in the document]
Comment (swghosh) [reply modules/cert-manager-enable-operator-log-level.adoc:22]: Good catch, @xingxingxia

**Comment by xenolinux** [comment]
@xingxingxia This PR is ready for the QE review.

File: modules/cert-manager-enable-operand-log-level.adoc [hunk 5]
```diff
@@ -0,0 +1,53 @@
+// Module included in the following assemblies:
+//
+// * security/cert_manager_operator/cert-manager-log-levels.adoc
+
+:_content-type: PROCEDURE
+[id="cert-manager-enable-operand-log-level_{context}"]
+= Setting a log level for cert-manager
+
+You can set a log level for cert-manager to determine the amount of detail in logged error messages.
```
Comment (xingxingxia) [inline modules/cert-manager-enable-operand-log-level.adoc:9]: This one also needs update like Swarup commented here https://github.com/openshift/openshift-docs/pull/59777#discussion_r1190691164 .
```suggestion
You can set a log level for cert-manager to determine the verbosity of log messages.
```

File: modules/cert-manager-enable-operand-log-level.adoc [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+metadata:
+  creationTimestamp: "2023-05-10T07:34:15Z"
+  finalizers:
+  - cert-manager-operator.operator.openshift.io/cert-manager-controller-deployment
+  - cert-manager-operator.operator.openshift.io/cert-manager-webhook-deployment
+  - cert-manager-operator.operator.openshift.io/cert-manager-cainjector-deployment
+  generation: 1
+  name: cluster
+  resourceVersion: "47290"
+  uid: 0f5cd145-a20f-43d1-ac3d-0db28a6f9f06
```
Comment (xingxingxia) [inline modules/cert-manager-enable-operand-log-level.adoc:40]: Fields creationTimestamp, resourceVersion, uid, et al ... are not needed to be shown.
```suggestion
apiVersion: operator.openshift.io/v1alpha1
kind: CertManager
...
```

File: modules/cert-manager-enable-operand-log-level.adoc [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+spec:
+  logLevel: Normal <1>
+----
+<1> Replace `<log_level>` with the desired log level value. The default `logLevel` is `Normal`. The valid log level values for the `CertManager` resource are `Normal`, `Debug`, `Trace`, and `TraceAll`. You can set `logLevel` to `Normal` to audit logs and perform common operations when everything is fine. To troubleshoot a minor issue, you can set `logLevel` to `Debug` to view verbose logs. To troubleshoot a major issue, you can set `logLevel` to `Trace` to view more verbose logs. To troubleshoot serious issues, you can set `logLevel` to `TraceAll`.
```
Comment (xingxingxia) [inline modules/cert-manager-enable-operand-log-level.adoc:44]: Didn't see `log_level` placeholder.
```suggestion
<1> Replace `Normal` with the desired log level value. The default `logLevel` is `Normal`. The valid log level values for the `CertManager` resource are `Normal`, `Debug`, `Trace`, and `TraceAll`. You can set `logLevel` to `Normal` to audit logs and perform common operations when everything is fine. To troubleshoot a minor issue, you can set `logLevel` to `Debug` to view verbose logs. To troubleshoot a major issue, you can set `logLevel` to `Trace` to view more verbose logs. To troubleshoot serious issues, you can set `logLevel` to `TraceAll`.
```

File: modules/cert-manager-enable-operand-log-level.adoc [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
++
+[NOTE]
+====
+`TraceAll` generates huge amount of logs. After setting `logLevel` to `TraceAll`, you might experience performance issues.
+====
+
+. Save the file to apply the changes.
```
Comment (xingxingxia) [inline modules/cert-manager-enable-operand-log-level.adoc:51]: It does oc edit to save changes for a resource, not usual sense of save a `file` (though oc edit uses temporary file)
```suggestion
. Save your changes and quit the text editor to apply your changes.
```

File: modules/cert-manager-enable-operator-log-level.adoc [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+----
+<1> Replace `v` with the desired log level number. The valid values for `v` are `1`, `2`, `3`, `4`, `5`, `6`, `7`, `8`, `9`, and `10`. The default value is `2`.
+
+.Verification
+
+* Verify that the log level of the {cert-manager-operator} is updated by running the following command:
++
+[source,terminal]
+----
+$ oc logs deploy/cert-manager-operator-controller-manager -n cert-manager-operator
```
Comment (xingxingxia) [inline modules/cert-manager-enable-operator-log-level.adoc:32]:
```suggestion
* The pod of {cert-manager-operator} will be redeployed. Verify that the log level of the {cert-manager-operator} is updated by running the following commands:
+
[source,terminal]
----
$ oc set env deploy/cert-manager-operator-controller-manager -n cert-manager-operator --list | grep -e OPERATOR_LOG_LEVEL -e container
# deployments/cert-manager-operator-controller-manager, container kube-rbac-proxy
OPERATOR_LOG_LEVEL=9
# deployments/cert-manager-operator-controller-manager, container cert-manager-operator
OPERATOR_LOG_LEVEL=9

$ oc logs deploy/cert-manager-operator-controller-manager -n cert-manager-operator
```

File: security/cert_manager_operator/cert-manager-log-levels.adoc [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+include::_attributes/common-attributes.adoc[]
+:context: cert-manager-log-levels
+
+toc::[]
+
+To troubleshoot some issues with the cert-manager components and the {cert-manager-operator}, you can configure the log level verbosity.
+
+[NOTE]
+====
+To use different log levels for different cert-manager components, see _Customizing cert-manager by using the cert-manager Operator API fields_.
```
Comment (xingxingxia) [inline security/cert_manager_operator/cert-manager-log-levels.adoc:13]: Need there be a link for `Customizing cert-manager by using the cert-manager Operator API fields`? (either in-place hypertext link or under "Additional resources")

**Comment by xingxingxia** [comment]
LGTM from QE side except above minor comment.

File: security/cert_manager_operator/cert-manager-log-levels.adoc (hunk 10 above)
Comment (xenolinux) [reply security/cert_manager_operator/cert-manager-log-levels.adoc:13]: @xingxingxia Yes. We will have to wait for the [PR](https://github.com/openshift/openshift-docs/pull/59726) to get merged. After that, I will add the section link under Additional Resources. If we add it now the `xref` will make the checks fail

Comment (xingxingxia) [reply security/cert_manager_operator/cert-manager-log-levels.adoc:13]: OK
/lgtm

File: modules/cert-manager-enable-operand-log-level.adoc [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+:_content-type: PROCEDURE
+[id="cert-manager-enable-operand-log-level_{context}"]
+= Setting a log level for cert-manager
+
+You can set a log level for cert-manager to determine the verbosity of log messages.
+
+.Prerequisites
+
+* You have access to the cluster with `cluster-admin` privileges.
+* The {cert-manager-operator} 1.11.1 or later is installed.
```
Comment (snarayan-redhat) [inline modules/cert-manager-enable-operand-log-level.adoc:14]: Just a suggestion that you can consider: Is it possible to make this active voice?
E.g.
```suggestion
* You have installed the {cert-manager-operator} 1.11.1 or later.
```
If not, that's fine.

File: modules/cert-manager-enable-operand-log-level.adoc [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
++
+[source,yaml]
+----
+apiVersion: operator.openshift.io/v1alpha1
+kind: CertManager
+...
+spec:
+  logLevel: Normal <1>
+----
+<1> The default `logLevel` is `Normal`. Replace `Normal` with the desired log level value. The valid log level values for the `CertManager` resource are `Normal`, `Debug`, `Trace`, and `TraceAll`. You can set `logLevel` to `Normal` to audit logs and perform common operations when everything is fine. To troubleshoot a minor issue, you can set `logLevel` to `Debug` to view verbose logs. To troubleshoot a major issue, you can set `logLevel` to `Trace` to view more verbose logs. To troubleshoot serious issues, you can set `logLevel` to `TraceAll`.
```
Comment (snarayan-redhat) [inline modules/cert-manager-enable-operand-log-level.adoc:35]:
```suggestion
<1> The default `logLevel` is `Normal`. Replace `Normal` with the desired log level value. The valid log level values for the `CertManager` resource are `Normal`, `Debug`, `Trace`, and `TraceAll`. You can set `logLevel` to `Normal` to audit logs and perform common operations when everything is fine. You can set `logLevel` to `Debug` to view verbose logs for troubleshooting a minor issue. You can set `logLevel` to `Trace` to view more verbose logs for troubleshooting a major issue. You can set `logLevel` to `TraceAll` for troubleshooting serious issues.
```

OR

```suggestion
<1> The default `logLevel` is `Normal`. Replace `Normal` with the desired log level value. The valid log level values for the `CertManager` resource are `Normal`, `Debug`, `Trace`, and `TraceAll`. To audit logs and perform common operations when everything is fine, set `logLevel` to `Normal` . To troubleshoot a minor issue by viewing verbose logs, set `logLevel` to `Debug` . To troubleshoot a major issue by viewing more verbose logs, you can set `logLevel` to `Trace`. To troubleshoot serious issues, set `logLevel` to `TraceAll`.
```

File: modules/cert-manager-enable-operator-log-level.adoc [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+
+.Procedure
+
+* Update the subscription object for {cert-manager-operator} to provide the verbosity level for the operator logs by running the following command:
++
+[source,terminal]
+----
+$ oc -n cert-manager-operator patch subscription openshift-cert-manager-operator --type='merge' -p '{"spec":{"config":{"env":[{"name":"OPERATOR_LOG_LEVEL","value":"v"}]}}}' <1>
+----
+<1> Replace `v` with the desired log level number. The valid values for `v` are `1`, `2`, `3`, `4`, `5`, `6`, `7`, `8`, `9`, and `10`. The default value is `2`.
```
Comment (snarayan-redhat) [inline modules/cert-manager-enable-operator-log-level.adoc:24]: Just a suggestion
```suggestion
<1> Replace `v` with the desired log level number. The valid values for `v` can range from `1`to `10`. The default value is `2`.
```

File: security/cert_manager_operator/cert-manager-log-levels.adoc [hunk 14]
```diff
@@ -0,0 +1,18 @@
+:_content-type: ASSEMBLY
+[id="cert-manager-log-levels"]
+= Configuring log levels for cert-manager and the {cert-manager-operator}
+include::_attributes/common-attributes.adoc[]
+:context: cert-manager-log-levels
+
+toc::[]
+
+To troubleshoot some issues with the cert-manager components and the {cert-manager-operator}, you can configure the log level verbosity.
```
Comment (snarayan-redhat) [inline security/cert_manager_operator/cert-manager-log-levels.adoc:9]: is `some` necessary? If yes, can `some` be defined here?

File: modules/cert-manager-enable-operator-log-level.adoc [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+:_content-type: PROCEDURE
+[id="cert-manager-enable-operator-log-level_{context}"]
+= Setting a log level for the {cert-manager-operator}
+
+You can set a log level for the {cert-manager-operator} to determine the verbosity of the operator log messages.
+
+.Prerequisites
+
+* You have access to the cluster with `cluster-admin` privileges.
+* The {cert-manager-operator} 1.11.1 or later is installed.
```
Comment (snarayan-redhat) [inline modules/cert-manager-enable-operator-log-level.adoc:14]: Same comment as the previous module.

**Review by snarayan-redhat — COMMENTED** [review]
Great job!! Few nits and suggestions.

File: security/cert_manager_operator/cert-manager-log-levels.adoc (hunk 14 above)
Comment (xenolinux) [reply security/cert_manager_operator/cert-manager-log-levels.adoc:9]: `some` can be avoided

**Pull request merged**